### ISSUE TITLE
Unexpected response from debug_traceTransaction

### ISSUE THREAD

**Issue by MinwooJ** [ethereum/go-ethereum#22846]
#### System information

Geth version: `v1.10.2`
OS & Version: Linux

#### Expected behaviour
When `debug_traceTransaction` with `{"tracer": "callTracer"}` option, I expected to get general response like actions.

#### Actual behaviour
It usually responses fine, but intermittently gives `{code=-32000, message=genesis is not traceable}` error Or `{code=-32000, message=transaction 0b3251312a4091994164b1de4fc10900527b656cc08ef3dd90a82d7921afff1c not found}` error

#### Steps to reproduce the behaviour
I am running a scheduler which `eth_getBlockByNumber`  first and `debug_traceTransaction` using the transaction hashes from the getBlock response.

**Comment by karalabe** [comment]
`{code=-32000, message=genesis is not traceable}` -> You tried to trace block `0`

Are you tracing old blocks? Do you have an archive node? If you're tracing historical blocks on a non-archive full node, you might want to look into debug.traceChain, which can stream traces from block to block without needing to regenerate stuff in between (won't work for newly mined blocks).

**Comment by MinwooJ** [comment]
@karalabe I am running the archive node and what I am curious about is usually it works well but intermittently it gives that error message. 

What I found is the error message occurs when the chain reorg happens. 

I runned the node fast sync at the first, and after the sync is completed, I changed it as full and archive. 

And this happened at Ropsten testnet.

**Comment by MariusVanDerWijden** [comment]
> What I found is the error message occurs when the chain reorg happens.

There's an inherent race condition when a reorg happens. You get the block `a` containing the old transactions, now a reorg happens and `a` is not part of the canonical chain anymore, this means that some of the transactions in `a` might not have been executed in the current head `b`. There's nothing to be done about that. You need to discard the traces of the transactions of `a` and instead retrieve the transactions from `b` and trace them.

**Comment by StevenProhashing** [comment]
I've noticed a similar issue happening for us. We have a system setup that uses debug_traceTransaction to check transactions sent from our system's wallet. Normally for pending unconfirmed transactions from our wallet, debug_traceTransaction would return a "not found" error message. But now it's returning the weird "genesis is not traceable" message. 

This first time we noticed this message was sometime in late April 2021. We upgraded to v1.10.2 in mid April for the Berlin fork. Before upgrading, we were running v1.9.25. We're running a linux non-archive full node.
